### PR TITLE
Add spring animation support

### PR DESCRIPTION
**Pull request by wieslawsoltes**
## What does the pull request do?
Adds spring animation support

## What is the current behavior?

Spring animation support is missing

## What is the updated/expected behavior with this PR?

You can add spring animation easing for animations

## How was the solution implemented (if it's not obvious)?

Used spring solver to implement spring object used in spring easing

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023922-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Animation/Spring.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+using Avalonia;
+using Avalonia.Utilities;
+
+namespace Avalonia.Animation;
+
+/// <summary>
+/// Determines how an animation is used based on spring formula.
+/// </summary>
+[TypeConverter(typeof(SpringTypeConverter))]
+public class Spring : AvaloniaObject
```
Comment (kekekeks) [inline src/Avalonia.Base/Animation/Spring.cs:13]: Are you sure that it should be AvaloniaObject?

Comment (wieslawsoltes) [reply src/Avalonia.Base/Animation/Spring.cs]: I am not sure, just based on https://github.com/AvaloniaUI/Avalonia/blob/d2146ba04e8336472e479e2159b0ea44eb249f86/src/Avalonia.Base/Animation/KeySpline.cs#L19-L20

Comment (wieslawsoltes) [reply src/Avalonia.Base/Animation/Spring.cs]: Remove AvaloniaObject usage

**Comment by wieslawsoltes** [comment]
@kekekeks Refactored spring solver so it can potentially be used in compositor `SpringScalarNaturalMotionAnimation` as it has different params:
- [Perdiod](https://learn.microsoft.com/en-us/uwp/api/windows.ui.composition.springscalarnaturalmotionanimation.period?view=winrt-19041#windows-ui-composition-springscalarnaturalmotionanimation-period)
- [DampingRatio ](https://learn.microsoft.com/en-us/uwp/api/windows.ui.composition.springscalarnaturalmotionanimation.dampingratio?view=winrt-19041)
- [InitialVelocity](https://learn.microsoft.com/en-us/uwp/api/windows.ui.composition.scalarnaturalmotionanimation.initialvelocity?view=winrt-19041#windows-ui-composition-scalarnaturalmotionanimation-initialvelocity)

https://github.com/AvaloniaUI/Avalonia/pull/8992/files#diff-c750554361b238a25705c35c16420177d96e7335ceaa1548b8bc7133307f156eR46

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023941-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

File: src/Avalonia.Base/Animation/Spring.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+using System.Globalization;
+using Avalonia;
+using Avalonia.Utilities;
+
+namespace Avalonia.Animation;
+
+/// <summary>
+/// Determines how an animation is used based on spring formula.
+/// </summary>
+[TypeConverter(typeof(SpringTypeConverter))]
```
Comment (kekekeks) [inline src/Avalonia.Base/Animation/Spring.cs:12]: Note that XAML compiler doesn't need a converter if type has a static Parse method

**Comment by kekekeks** [comment]
Are `Spring` parameters based on some web standard?

**Comment by wieslawsoltes** [comment]
> Are `Spring` parameters based on some web standard?

Based on WebKit source code and laws of physics, other solution is to use compositor parameters like period and damping ratio instead of mass and stiffness. 

https://github.com/WebKit/WebKit/blob/8afe31a018b11741abdf9b4d5bb973d7c1d9ff05/Websites/webkit.org/demos/spring/spring.js#L26
https://github.com/WebKit/WebKit/blob/4cc03269dc5d1b1e555e47ba70fbfccb31835802/Source/WebCore/platform/animation/TimingFunction.cpp#L83
https://trac.webkit.org/changeset/201759/webkit
https://codepen.io/shospodarets/pen/ZebBvV?editors=0100

**Comment by wieslawsoltes** [comment]
> > Are `Spring` parameters based on some web standard?
> 
> Based on WebKit source code and laws of physics, other solution is to use compositor parameters like period and damping ratio instead of mass and stiffness.
> 
> [WebKit/WebKit@`8afe31a`/Websites/webkit.org/demos/spring/spring.js#L26](https://github.com/WebKit/WebKit/blob/8afe31a018b11741abdf9b4d5bb973d7c1d9ff05/Websites/webkit.org/demos/spring/spring.js#L26) [WebKit/WebKit@`4cc0326`/Source/WebCore/platform/animation/TimingFunction.cpp#L83](https://github.com/WebKit/WebKit/blob/4cc03269dc5d1b1e555e47ba70fbfccb31835802/Source/WebCore/platform/animation/TimingFunction.cpp#L83) [trac.webkit.org/changeset/201759/webkit](https://trac.webkit.org/changeset/201759/webkit) [codepen.io/shospodarets/pen/ZebBvV?editors=0100](https://codepen.io/shospodarets/pen/ZebBvV?editors=0100)

Don't see the spring in standard https://www.w3schools.com/cssref/css3_pr_transition-timing-function.asp

**Comment by wieslawsoltes** [comment]
> Are `Spring` parameters based on some web standard?

It's also inspired how Figma does spring animation https://www.figma.com/blog/how-we-built-spring-animations/ (they used same WebKit solver as I in this PR).

**Review by kekekeks — APPROVED** [review]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0023959-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

**Pull request merged**